src/Microsoft.SqlTools.ServiceLayer/Profiler/ProfilerSessionMonitor.cs

```diff
@@ -50,35 +72,111 @@ public void AddSessionListener(IProfilerSessionListener listener)
+                    this.processorThread = Task.Factory.StartNew(ProcessSessions);
+                }
+
+                // create new profiling session if needed
+                if (!this.monitoredSessions.ContainsKey(session.ID))
+                {
+                    var profilerSession = new ProfilerSession();
+                    profilerSession.XEventSession = session;
+
+                    this.monitoredSessions.Add(session.ID, profilerSession);
                 }
 
-                if (!this.monitoredSessions.ContainsKey(session.SessionId))
+                // create a new viewer, or configure existing viewer
+                Viewer viewer;
+                if(!this.allViewers.TryGetValue(viewerID, out viewer))
                 {
-                    this.monitoredSessions.Add(session.SessionId, session);
+                    viewer = new Viewer(viewerID, true, session.ID);
+                    allViewers.Add(viewerID, viewer);
+                }
+                else
+                {
+                    viewer.active = true;
+                    viewer.xeSessionID = session.ID;
+                }
+
+                // add viewer to XEvent session viewers
+                List<string> viewers;
+                if(this.sessionViewers.TryGetValue(session.ID, out viewers))
+                {
+                    viewers.Add(viewerID);
+                }
+                else
+                {
+                    viewers = new List<string>{ viewerID };
+                    sessionViewers.Add(session.ID, viewers);
                 }
             }
 
             return true;
         }
 
+        // TODO: Clean this up
+        /// <summary>
+        /// Stop monitoring the session watched by viewerID
+        /// </summary>
+        public bool StopMonitoringSession(string viewerID, out ProfilerSession session)
+        {
+            lock (this.sessionsLock)
+            {
+                Viewer v;
+                if(this.allViewers.TryGetValue(viewerID, out v))
+                {
+                    return RemoveSession(v.xeSessionID, out session);
+                }
+                else
+                {
+                    session = null;
+                    return false;
+                }
+            }
+        }
+
         /// <summary>
-        /// Stop monitoring the session specified by the sessionId
+        /// Toggle the pause state for the viewer
         /// </summary>
-        public bool StopMonitoringSession(string sessionId, out ProfilerSession session)
+        public void PauseViewer(string viewerID)
         {
+
             lock (this.sessionsLock)
             {
-                if (this.monitoredSessions.ContainsKey(sessionId))
+                //update the status in all viewers
+                Viewer v = this.allViewers[viewerID];
+                this.allViewers[viewerID] = new Viewer(viewerID, !v.active, v.xeSessionID);
```

Comment:
Is this just updating `v`'s `active` property? Why not do it directly instead of creating a new `Viewer`?